// @ngrx/devtools/dist/monitors/devtools.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var log_monitor_1 = require('./log-monitor/log-monitor');
var dock_monitor_1 = require('./dock-monitor/dock-monitor');
var Devtools = (function () {
    function Devtools() {
        this.toggleCommand = 'ctrl-h';
        this.positionCommand = 'ctrl-m';
    }
    __decorate([
        core_1.Input('toggle-command'), 
        __metadata('design:type', String)
    ], Devtools.prototype, "toggleCommand", void 0);
    __decorate([
        core_1.Input('position-command'), 
        __metadata('design:type', String)
    ], Devtools.prototype, "positionCommand", void 0);
    Devtools = __decorate([
        core_1.Component({
            selector: 'ngrx-devtools',
            directives: [log_monitor_1.LogMonitor, dock_monitor_1.DockMonitor],
            template: "\n    <dock-monitor [toggleCommand]=\"toggleCommand\" [positionCommand]=\"positionCommand\">\n      <log-monitor></log-monitor>\n    </dock-monitor>\n  "
        }), 
        __metadata('design:paramtypes', [])
    ], Devtools);
    return Devtools;
}());
exports.Devtools = Devtools;
