// @ngrx/devtools/dist/monitors/log-monitor/log-monitor-entry.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var log_monitor_button_1 = require('./log-monitor-button');
var LogMonitorEntry = (function () {
    function LogMonitorEntry() {
        this.toggle = new core_1.EventEmitter();
    }
    LogMonitorEntry.prototype.handleToggle = function () {
        this.toggle.next({ id: this.item.actionId });
    };
    LogMonitorEntry.prototype.logPayload = function () {
        console.log(this.item.action);
    };
    LogMonitorEntry.prototype.logState = function () {
        console.log(this.item.state);
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], LogMonitorEntry.prototype, "item", void 0);
    __decorate([
        core_1.Output(), 
        __metadata('design:type', Object)
    ], LogMonitorEntry.prototype, "toggle", void 0);
    __decorate([
        core_1.HostListener('click'), 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', []), 
        __metadata('design:returntype', void 0)
    ], LogMonitorEntry.prototype, "handleToggle", null);
    LogMonitorEntry = __decorate([
        core_1.Component({
            selector: 'log-monitor-entry',
            changeDetection: core_1.ChangeDetectionStrategy.OnPush,
            directives: [log_monitor_button_1.LogMonitorButton],
            encapsulation: core_1.ViewEncapsulation.Emulated,
            template: "\n    <div class=\"title-bar\" [ngClass]=\"{ collapsed: item.collapsed }\">\n      {{ item.action.type }}\n    </div>\n    <div class=\"action-bar\" *ngIf=\"!item.collapsed\">\n      <log-monitor-button (action)=\"logPayload($event)\">\n        Log Payload\n      </log-monitor-button>\n      <log-monitor-button (action)=\"logState($event)\">\n        Log State\n      </log-monitor-button>\n    </div>\n  ",
            styles: ["\n    :host{\n      color: #FFFFFF;\n      background-color: #4F5A65;\n      cursor: pointer;\n    }\n    .title-bar{\n      padding: 8px 0 7px 16px;\n      background-color: rgba(0,0,0,0.1);\n    }\n    .action-bar{\n      padding: 8px 0 7px;\n      text-align: center;\n    }\n    .collapsed{\n      text-decoration: line-through;\n      font-style: italic;\n      opacity: 0.5;\n    }\n    log-monitor-button{\n      opacity: 0.6;\n    }\n  "]
        }), 
        __metadata('design:paramtypes', [])
    ], LogMonitorEntry);
    return LogMonitorEntry;
}());
exports.LogMonitorEntry = LogMonitorEntry;
